File: JavaTest/src/main/java/test/StringBuilderTest.java
package test;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

/**
 * The {@code StringBuilderTest} class performs automatic testing of the
 * {@code StringBuilder} class.
 *
 * @author Petr Kozler
 */
public class StringBuilderTest {

    public static Logger logger = LogManager.getLogger(StringBuilderTest.class.getName());

    public static void runTestSuite() {
        StringBuilderTest.stringBuilderTest();
        StringBuilderTest.appendTest();
        StringBuilderTest.toStringTest();
    }

    /**
     * Tests the constructor.
     */
    public static void stringBuilderTest() {
        StringBuilder a;
        String b;
        // new empty stringbuilder
        a = new StringBuilder();
        b = a.toString();
        logger.info(b);
    }

    /**
     * Tests the append method.
     */
    public static void appendTest() {
        StringBuilder a;
        String b;
        // appending of empty string
        a = new StringBuilder();
        a.append("");
        b = a.toString();
        logger.info(b);
        // appending of one char
        a = new StringBuilder();
        a.append(".");
        b = a.toString();
        logger.info(b);
        // appending of more chars
        a = new StringBuilder();
        a.append("test");
        b = a.toString();
        logger.info(b);
        // multiple appendings (empty string at the beginning)
        a = new StringBuilder();
        a.append("test");
        a.append(".");
        a.append("");
        b = a.toString();
        logger.info(b);
        // multiple appendings (empty string in the middle)
        a = new StringBuilder();
        a.append(".");
        a.append("");
        a.append("test");
        b = a.toString();
        logger.info(b);
        // multiple appendings (empty string at the end)
        a = new StringBuilder();
        a.append("");
        a.append("test");
        a.append(".");
        b = a.toString();
        logger.info(b);
    }

    /**
     * Tests the toString method.
     */
    public static void toStringTest() {
        StringBuilder a;
        String b;
        // appending a string at once
        a = new StringBuilder();
        a.append("This is a test");
        b = a.toString();
        logger.info(b);
        // appending a string piecewise
        a = new StringBuilder();
        a.append("This").append(" is").append(" a").append(" test.");
        b = a.toString();
        logger.info(b);
        // appending a string piecewise (more pieces)
        a = new StringBuilder();
        a.append("This").append(" ").append("is").append(" ").append("a").append(" ").append("test").append(".");
        b = a.toString();
        logger.info(b);
    }

}

File: C/CLibrary/tests/StringBuilderTest.c
#include <stdio.h>
#include <stdlib.h>

#include "test.h"
#include "StringBuilder.h"

/*
 * Simple C Test Suite
 */

/**
 * Tests the constructor.
 */

void stringBuilderTest() {
    StringBuilder *a;
    String *b;
    // new_empty stringbuilder
    a = new_StringBuilder();
    b = toStringSb(a);
    assertEqualsStr("", b->s);
}

/**
 * Tests the append function.
 */

void appendTest() {
    StringBuilder *a;
    String *b;
    // appending of empty string
    a = new_StringBuilder();
    append(a, new_String(""));
    b = toStringSb(a);
    assertEqualsStr("", b->s);
    // appending of one char
    a = new_StringBuilder();
    append(a, new_String("."));
    b = toStringSb(a);
    assertEqualsStr(".", b->s);
    // appending of more chars
    a = new_StringBuilder();
    append(a, new_String("test"));
    b = toStringSb(a);
    assertEqualsStr("test", b->s);
    // multiple appendings (empty string at the beginning)
    a = new_StringBuilder();
    append(a, new_String("test"));
    append(a, new_String("."));
    append(a, new_String(""));
    b = toStringSb(a);
    assertEqualsStr("test.", b->s);
    // multiple appendings (empty string in the middle)
    a = new_StringBuilder();
    append(a, new_String("."));
    append(a, new_String(""));
    append(a, new_String("test"));
    b = toStringSb(a);
    assertEqualsStr(".test", b->s);
    // multiple appendings (empty string at the end)
    a = new_StringBuilder();
    append(a, new_String(""));
    append(a, new_String("test"));
    append(a, new_String("."));
    b = toStringSb(a);
    assertEqualsStr("test.", b->s);
}

/**
 * Tests the toString function.
 */

void toStringTest() {
    StringBuilder *a;
    String *b;
    // appending a string at once
    a = new_StringBuilder();
    append(a, new_String("This is a test"));
    b = toStringSb(a);
    assertEqualsStr("This is a test", b->s);
    // appending a string piecewise
    a = new_StringBuilder();
    append(append(append(append(a, 
            new_String("This")), 
            new_String(" is")), 
            new_String(" a")), 
            new_String(" test."));
    b = toStringSb(a);
    assertEqualsStr("This is a test.", b->s);
    // appending a string piecewise (more pieces)
    a = new_StringBuilder();
    append(append(append(append(append(append(append(append(a, 
            new_String("This")), 
            new_String(" ")), 
            new_String("is")), 
            new_String(" ")), 
            new_String("a")), 
            new_String(" ")), 
            new_String("test")), 
            new_String("."));
    b = toStringSb(a);
    assertEqualsStr("This is a test.", b->s);
}

int main(int argc, char** argv) {
    const int testCount = 3;
    setUpTestModule("StringBuilderTest", testCount);

    registerTest(stringBuilderTest, "stringBuilderTest");
    registerTest(appendTest, "appendTest");
    registerTest(toStringTest, "toStringTest");
    
    runTests();

    return (EXIT_SUCCESS);
}
